// @bigbighu/altas/extends/logger.js
const { createLogger, format, transports } = require('winston');
const { combine, timestamp, printf } = format;
const customFormat = printf((info) => {
  let { message } = info;
  if (typeof message === 'object') {
    let cache = [];
    message = JSON.stringify(message, (key, value) => {
      if (typeof value === 'object' && value !== null) {
        if (cache.indexOf(value) !== -1) {
          try {
            return JSON.parse(JSON.stringify(value));
          } catch (error) {
            return undefined;
          }
        }
        cache.push(value);
      }
      return value;
    });
    cache = [];
  }
  return `[${info.timestamp}] [${info.level}] ${message}`;
});

const logger = createLogger({
  format: combine(
    format.colorize(),
    timestamp({
      format: 'YYYY-MM-DD HH:mm:ss',
    }),
    customFormat,

  ),
  transports: [
    new transports.File({
      name: 'info',
      filename: __dirname + '/../../../logs/debug.log',
      json: true,
      level:'info',
      maxsize: 1024 * 1024 * 10 // 10MB
    }),
    new transports.File({
      name: 'error',
      filename: __dirname + '/../../../logs/error.log',
      json: true,
      level:'error',
      maxsize: 1024 * 1024 * 10 // 10MB
    }),
    new transports.Console()
  ],
});

module.exports = () => {
  return (req, res, next) => {
    logger.info(`handle request url ${req.url}`);
    next();
  }
};
